src/main/kotlin/org/wycliffeassociates/otter/jvm/recorder/app/view/CanvasFragment.kt

```diff
@@ -0,0 +1,54 @@
+package org.wycliffeassociates.otter.jvm.recorder.app.view
+
+import javafx.geometry.Insets
+import javafx.geometry.Pos
+import javafx.scene.layout.*
+import javafx.scene.paint.Paint
+import org.wycliffeassociates.otter.jvm.recorder.app.view.drawables.Drawable
+import tornadofx.add
+import tornadofx.canvas
+import tornadofx.hgrow
+import tornadofx.vgrow
+
+
+class CanvasFragment(color: String = "#000000") : StackPane() {
+
+    private val drawables = arrayListOf<Drawable>()
+
+    private val cvs = canvas {
+        hgrow = Priority.ALWAYS
+        vgrow = Priority.ALWAYS
+    }
+    private val ctx = cvs.graphicsContext2D
+
+    init {
+        background = Background(BackgroundFill(Paint.valueOf(color), CornerRadii.EMPTY, Insets.EMPTY))
+
+        alignment = Pos.TOP_LEFT
+
+        add(cvs)
+
+        cvs.widthProperty().bind(this.widthProperty())
+        cvs.heightProperty().bind(this.heightProperty())
+
+        cvs.widthProperty().addListener { _ -> draw() }
+        cvs.heightProperty().addListener { _ -> draw() }
+    }
+
+    init {
+        draw()
+    }
+
+    fun addDrawable(drawable: Drawable) {
+        drawables.add(drawable)
+    }
+
+    fun draw() {
+        ctx.clearRect(0.0, 0.0, cvs.width, cvs.height)
+        var i = 0
+        while(i < drawables.size) {
```

Comment:
can this be converted to a foreach?

Reply:
No, a for each will allocate an iterable object that needs to be garbage collected, which causes a lot of GC runs since this is a draw loop